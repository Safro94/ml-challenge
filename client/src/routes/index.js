import React from 'react';
import { Route, Switch } from 'react-router-dom';

import routes from './routes';

export default () => (
	<Switch>
		{routes.map(route => (
			<Route key={route.path} {...route} />
		))}
	</Switch>
);
