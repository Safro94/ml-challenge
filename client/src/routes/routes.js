import loadable from '@loadable/component';

export const HOME = '/';
export const RESULTS = '/items';
export const DETAIL = '/items/:id';

const Home = loadable(() =>
	import(
		/* webpackChunkName: 'Home'  */
		'../pages/home'
	)
);

const Results = loadable(() =>
	import(
		/* webpackChunkName: 'Results'  */
		'../pages/results'
	)
);

const Detail = loadable(() =>
	import(
		/* webpackChunkName: 'Detail'  */
		'../pages/detail'
	)
);

export default [
	{
		path: HOME,
		component: Home,
		exact: true,
	},
	{
		path: DETAIL,
		component: Detail,
	},
	{
		path: RESULTS,
		component: Results,
	},
];
