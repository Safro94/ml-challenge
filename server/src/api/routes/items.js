const express = require('express');
const { getItems, getItemById } = require('../controllers/items');

const router = express.Router();

router.get('/', getItems);
router.get('/:id', getItemById);

module.exports = router;
