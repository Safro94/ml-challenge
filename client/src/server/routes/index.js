import express from 'express';
import React from 'react';
import path from 'path';

import { renderToString } from 'react-dom/server';
import { StaticRouter, matchPath } from 'react-router-dom';

import { ChunkExtractor, ChunkExtractorManager } from '@loadable/server';
import StyleContext from 'isomorphic-style-loader/StyleContext';

import routes from 'routes/routes';

import App from 'components/app';

import template from '../template';

const server = express();

server
	.disable('x-powered-by')
	.use(express.static(process.env.RAZZLE_PUBLIC_DIR))
	.get('/*', async (req, res) => {
		/*
			Obtengo la ruta de la url y matcheo con mis rutas
			para ver si tiene un requestInitialData y fetchear
			la data en el server asi viene la vista armada. Si
			no tiene requestInitialData, devuelvo un objeto vacio
		*/
		const currentRoute = routes.find(route => matchPath(req.path, route));
		let initialData;
		if (currentRoute?.requestInitialData) {
			initialData = await currentRoute.requestInitialData(req);
		} else {
			initialData = {};
		}
		const context = { initialData };

		const css = new Set();
		const insertCss = (...styles) =>
			styles.forEach(style => css.add(style._getCss()));

		const extractor = new ChunkExtractor({
			statsFile: path.resolve('build/loadable-stats.json'),
			entrypoints: ['client'],
		});

		const markup = renderToString(
			<ChunkExtractorManager extractor={extractor}>
				<StaticRouter context={context} location={req.url}>
					<StyleContext.Provider value={{ insertCss }}>
						<App />
					</StyleContext.Provider>
				</StaticRouter>
			</ChunkExtractorManager>
		);

		res.send(
			template({
				markup,
				js: extractor,
				styles: css,
				initialData,
			})
		);
	});

export default server;
