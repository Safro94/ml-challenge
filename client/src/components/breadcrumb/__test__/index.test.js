import React from 'react';

import { render } from '@testing-library/react';
import { screen } from '@testing-library/dom';
import userEvent from '@testing-library/user-event';

import { Link } from 'react-router-dom';

import { HOME } from 'routes/routes';

import Breadcrumb from '../';

jest.mock('react-router-dom', () => ({
	Link: jest.fn(({ children, ...rest }) => <a {...rest}>{children}</a>),
}));

describe('Breadcrumb', () => {
	const category = 'category';

	it('should render OK when isLast is false', () => {
		//Act
		const { container } = render(
			<Breadcrumb category={category} isLast={false} />
		);

		//Assert
		expect(container).toMatchSnapshot();
	});

	it('should render OK when isLast is true', () => {
		//Act
		const { container } = render(<Breadcrumb category={category} isLast />);

		//Assert
		expect(container).toMatchSnapshot();
	});

	it('should have a > when it is not the last item', () => {
		//Act
		render(<Breadcrumb category={category} isLast={false} />);

		//Assert
		expect(screen.getByText(/>/)).toBeInTheDocument();
	});

	it('should not have a > when it is the last item', () => {
		//Act
		render(<Breadcrumb category={category} isLast />);

		//Assert
		expect(screen.queryByText(/>/)).not.toBeInTheDocument();
	});

	it('should go to the path when the link is clicked', () => {
		//Arrange
		const path = '/path';

		//Act
		render(<Breadcrumb category={category} path={path} />);
		userEvent.click(screen.getByTestId('link'));

		//Assert
		expect(Link).toHaveBeenCalledTimes(1);
		expect(Link.mock.calls[0][0].to).toBe(path);
	});

	it('should go to the home page when the link is clicked and there is no path', () => {
		//Act
		render(<Breadcrumb category={category} />);
		userEvent.click(screen.getByTestId('link'));

		//Assert
		expect(Link).toHaveBeenCalledTimes(1);
		expect(Link.mock.calls[0][0].to).toBe(HOME);
	});
});
