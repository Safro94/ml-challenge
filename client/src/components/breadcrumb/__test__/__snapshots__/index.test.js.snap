// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Breadcrumb should render OK when isLast is false 1`] = `
<div>
  <li
    class="breadcrumb"
  >
    <a
      class="link"
      data-testid="link"
      to="/"
    >
      category
    </a>
    <span
      class="separator"
    >
       
      &gt;
       
    </span>
  </li>
</div>
`;

exports[`Breadcrumb should render OK when isLast is true 1`] = `
<div>
  <li
    class="breadcrumb"
  >
    <a
      class="link last"
      data-testid="link"
      to="/"
    >
      category
    </a>
  </li>
</div>
`;
